aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -586,11 +720,33 @@ void scatter_fill_cpu_kernel(const Tensor& self, int64_t dim, const Tensor& inde
     self, dim, index, value, "scatter_fill_cpu_", tensor_assign);
 }
 
-void scatter_add_cpu_kernel(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& src) {
-  cpu_scatter_gather_base_kernel<>()(
-    self, dim, index, src,
-    "scatter_add_", reduce_add);
+inline bool with_spmm_for_scatter(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& src) {
+  if (!is_radix_sort_available()) { return false; }
+
+  // skip when having empty tensor
+  if (self.numel() == 0 || index.numel() == 0 || src.numel() == 0) { return false; }
+
+  // skip when having scalar tensor
+  if (self.ndimension() == 0 || index.ndimension() == 0 || src.ndimension() == 0) { return false; }
 
+  // using `spmm` for scatter would require sorting on index,
+  // this is only perf beneficial when the inner dimension, aka, `channels`
+  // is big enough.
+  constexpr int64_t threshold = 16;
+  if (index.numel() / index.size(0) < threshold) { return false; }
+
+  // index is expanded
+  return dim == 0 && index.stride(0) == 1 && src.is_contiguous() && self.is_contiguous();
```

Comment:
Is this check too loose for an expanded tensor? I guess you should also check the strides of other dims are all zeros?

Reply:
the actual use case might have strides like `[1, 0, 0]`, `[1, 0, 1]` for 3d tensors. It should be OK if dim0 has stride of 1 (also this will make sure that index is contiguous).

Reply:
Oh, my concern is that `index.stride(0) == 1` is not sufficient to decide the `index` is expanded or not. It could be a non-contiguous tensor with `stride(0) == 1` but not expanded from 1d, e.g.,
```bash
>>> import torch
>>> index = torch.randint(0, 100, [50, 40, 30])
>>> index = index.transpose(0,2).contiguous().transpose(0,2)
>>> index.stride()
(1, 50, 2000)
```
Would this be a problem?

Reply:
Good point! Updated!